// ng-packagr/lib/flatten/file-loader-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fileLoaderPlugin = fileLoaderPlugin;
const node_path_1 = require("node:path");
const log = __importStar(require("../utils/log"));
const path_1 = require("../utils/path");
const POTENTIAL_MATCHES = ['', '.mjs', '/index.mjs'];
/**
 * Loads a file and it's map.
 */
function fileLoaderPlugin(fileCache) {
    return {
        name: 'file-loader',
        resolveId: function (id, importer) {
            if (fileCache.has((0, path_1.ensureUnixPath)(id))) {
                return id;
            }
            if (importer) {
                const resolved = (0, path_1.ensureUnixPath)((0, node_path_1.resolve)((0, node_path_1.dirname)(importer), id));
                for (const suffix of POTENTIAL_MATCHES) {
                    const potential = resolved + suffix;
                    if (fileCache.has(potential)) {
                        return potential;
                    }
                }
            }
        },
        load: function (id) {
            log.debug(`file-loader ${id}`);
            const data = fileCache.get((0, path_1.ensureUnixPath)(id));
            if (!data) {
                throw new Error(`Could not load '${id}' from memory.`);
            }
            return {
                code: data.content,
                map: data.map,
            };
        },
    };
}
